tokio/src/runtime/tests/handle.rs

```diff
@@ -0,0 +1,296 @@
+
+#[test]
+fn current_thread_block_on_with_fs() {
+    let rt = new_current_thread();
+    let _enter = rt.enter();
+    let handle = rt.handle();
+
+    let contents = handle.block_on(fs::read_to_string("Cargo.toml")).unwrap();
+
+    assert!(contents.contains("tokio"));
+}
+
+// ==== current threaded block_on after shutdown
+
+#[test]
+fn current_thread_block_on_executor_independent_future_with_dropped_runtime_works() {
+    let rt = new_current_thread();
+    let _enter = rt.enter();
+    let handle = rt.block_on(async { Handle::current() });
+
+    rt.shutdown_timeout(Duration::from_secs(1000));
+
+    let one = handle.block_on(async { 1 });
+    assert_eq!(1, one);
+}
+
+#[test]
+#[should_panic(expected = "Cannot run timers on a runtime that has been shutdown")]
+fn current_thread_block_on_with_timer_with_dropped_runtime_panics() {
+    let rt = new_current_thread();
+    let _enter = rt.enter();
+    let handle = rt.block_on(async { Handle::current() });
+
+    rt.shutdown_timeout(Duration::from_secs(1000));
+
+    handle.block_on(time::sleep(Duration::from_millis(10)));
+}
+
+#[test]
+#[should_panic(expected = "Cannot run timers on a runtime that has been shutdown")]
+fn current_thread_block_on_with_timer_created_before_shutdown_with_dropped_runtime_panics() {
+    let rt = new_current_thread();
+    let _enter = rt.enter();
+    let handle = rt.block_on(async { Handle::current() });
+
+    let timer = time::sleep(Duration::from_millis(10));
+
+    rt.shutdown_timeout(Duration::from_secs(1000));
+
+    handle.block_on(timer);
+}
+
+#[test]
+#[should_panic(expected = "Cannot run I/O operations on a runtime that has been shutdown")]
+fn current_thread_block_on_with_fs_with_dropped_runtime_panics() {
+    let rt = new_current_thread();
+    let _enter = rt.enter();
+    let handle = rt.block_on(async { Handle::current() });
+
+    rt.shutdown_timeout(Duration::from_secs(1000));
+
+    let _ = handle.block_on(fs::read_to_string("Cargo.toml"));
+}
+
+// NOTE: currently fails
+#[test]
+#[ignore]
+#[should_panic(expected = "some message about not being to do io from dropped runtime")]
+fn current_thread_block_on_with_net_with_dropped_runtime_panics() {
+    let rt = new_current_thread();
+    let _enter = rt.enter();
+    let handle = rt.block_on(async { Handle::current() });
+
+    rt.shutdown_timeout(Duration::from_secs(1000));
+
+    let _ = handle.block_on(TcpStream::connect("127.0.0.1:0"));
+}
+
+// NOTE: currently fails
+#[test]
+#[ignore]
+#[should_panic(expected = "some message about not being to do io from dropped runtime")]
+fn current_thread_block_on_with_net_created_before_shutdown_with_dropped_runtime_panics() {
+    let rt = new_current_thread();
+    let _enter = rt.enter();
+    let handle = rt.block_on(async { Handle::current() });
+
+    let net_future = TcpStream::connect("127.0.0.1:0");
+
+    rt.shutdown_timeout(Duration::from_secs(1000));
+
+    let _ = handle.block_on(net_future);
```

Comment:
Should this panic? It might just need to return an IO error.

Reply:
It shouldn't panic to be consistent with timers and file io?